### DISCUSSION TITLE
Optional .idl Tag Error

### DISCUSSION THREAD

**Discussion by masu7056** [OpenDDS/OpenDDS#4832] [Q&A]
Good morning everyone,
I am trying to utilize the new `@optional` tag support provided in version 3.29+ and I keep encountering an error within the generation of the TypeSupportImpl.cpp file. I'm using OpenDDS 3.29.1 currently when I last received this error. The error specifies that my variable with the **Optional** Tag tied to it, like `@optional float var1` cannot be accessed as a function in the .cpp generated file. Error reads as shown: "**MemberParam param(0, false, "var1", 1, value.var1().has_value());** error: expression cannot be used as a function". Additionally, there are quite a few other errors regarding the same variable but in other uses, such as the Impl .cpp using value() as a method and the application sees it as a float. 

Is there any example that uses an .idl with a **Optional** tag applied that I could view or any solution to this? 

Thanks!

**Comment by jwillemsen** [comment]
See https://github.com/OpenDDS/OpenDDS/tree/master/tests/DCPS/Compiler/xcdr

**Comment by mitza-oci** [comment]
Did you see the note about using the IDL-to-C++11 language mapping here?
https://opendds.readthedocs.io/en/latest-release/devguide/xtypes.html#optional

If that doesn't help, please post complete and detailed steps to reproduce the issue with the current release of OpenDDS.

**Comment by masu7056** [comment]
Hi all,
Thank you for the quick responses and attempts to point me in the right direction. I am still struggling with the issue I've encountered earlier after some more attempts. Attached is my example .idl file that has an example variable that's giving errors. 
[StockQuoter.txt](https://github.com/user-attachments/files/17439394/StockQuoter.txt)
I had to adjust it to a .txt file, maybe I'm not creating my @optional annotation correctly..... 

My setup is as follows:

- Using OpenDDS 3.30
- configured with --ace-tao=ace8tao4 and -std=c++17
- In OpenDDS-DDS-3.30/examples/DCPS/IntroductionToOpenDDS
- Injected into StockQuoter.idl the above struct with @optional annotation
- ran $ACE_ROOT/bin/mwc.pl -type gnuace StockQuoter.mwc 
- then ran make

**Comment by sonndinh** [comment]
I don't see anything incorrect with how you use the `@optional` annotation in the modified IDL file. However, the StockQuoter example would need more changes to its source code and mpc file to make it work with `@optional`. The link above (https://github.com/OpenDDS/OpenDDS/issues/4824#issuecomment-2422901667) mentions the related changes.

For an example that already works, the Compiler/xcdr test (https://github.com/OpenDDS/OpenDDS/issues/4824#issuecomment-2421716422) is a good start. Note the use of `opendds_cxx11` in the xcdr.mpc file to generate code using C++11 mapping. `optional.idl` uses the `@optional` annotation.
